src/rgw/rgw_acl.cc

```diff
@@ -73,49 +70,41 @@ bool operator!=(const RGWAccessControlPolicy& lhs,
 void RGWAccessControlList::register_grant(const ACLGrant& grant)
 {
   ACLPermission perm = grant.get_permission();
-  ACLGranteeType type = grant.get_type();
-  switch (type.get_type()) {
-  case ACL_TYPE_REFERER:
-    referer_list.emplace_back(grant.get_referer(), perm.get_permissions());
+
+  if (const auto* user = grant.get_user(); user) {
+    acl_user_map[user->id.to_str()] |= perm.get_permissions();
+  } else if (const auto* email = grant.get_email(); email) {
+    acl_user_map[email->address] |= perm.get_permissions();
+  } else if (const auto* group = grant.get_group(); group) {
+    acl_group_map[group->type] |= perm.get_permissions();
+  } else if (const auto* referer = grant.get_referer(); referer) {
+    referer_list.emplace_back(referer->url_spec, perm.get_permissions());
 
     /* We're specially handling the Swift's .r:* as the S3 API has a similar
      * concept and thus we can have a small portion of compatibility here. */
-     if (grant.get_referer() == RGW_REFERER_WILDCARD) {
+     if (referer->url_spec == RGW_REFERER_WILDCARD) {
        acl_group_map[ACL_GROUP_ALL_USERS] |= perm.get_permissions();
      }
-    break;
-  case ACL_TYPE_GROUP:
-    acl_group_map[grant.get_group()] |= perm.get_permissions();
-    break;
-  default:
-    {
-      rgw_user id;
-      grant.get_id(id);
-      acl_user_map[id.to_str()] |= perm.get_permissions();
-    }
   }
 }
 
 void RGWAccessControlList::add_grant(const ACLGrant& grant)
 {
-  rgw_user id;
-  grant.get_id(id); // note that this will return false for groups, but that's ok, we won't search groups
-  grant_map.emplace(id.to_str(), grant);
+  std::string id;
+  if (const auto* user = grant.get_user(); user) {
```

Comment:
why `auto*` ?

Reply:
you're right that the `*` isn't required, but i figured it helps make it clear that `user` is a pointer that may be null

i see that i wasn't consistent about this, some places omit the *. overall, do you think the * hurts readability or helps? i'll update everything accordingly

Reply:
i wasn't familiar with the syntax so this was confusing, but overall it provides an extra level of type checking, so i would say, lets be consistent and use it